File: pi.go
package main

import "runtime"
import "time"
import "fmt"
import "os"
import "strconv"

var notify = make(chan int)
var result = make(chan float64)

func calc(id int, step float64, split int) {
    var x, pi float64
    base := id * split
    end := base + split

    <- notify
    fmt.Printf("%d start\n", id)

    for i := base; i < end; i++ {
        x = (float64(i) + 0.5) * step
        pi += 4 / (1 + x * x)
    }

    result <- pi
}

func main() {
    split, err := strconv.Atoi(os.Args[1])
    if err != nil {
        os.Exit(2)
    }
    num_thread := runtime.NumCPU()
    step := 1.0 / float64(split)
    thread_split := split / num_thread

    fmt.Printf("split: %d, step: %f, thread_split: %d\n", split, step, thread_split)

    start_t := time.Now()
    for i := 0; i < num_thread; i++ {
        go calc(i, step, thread_split)
        notify <- 0
    }

    var pi float64
    for i := 0; i < num_thread; i++ {
        pi += <-result
    }
    pi *= step
    elapsed := time.Since(start_t)

    fmt.Printf("pi: %f, time: %s\n", pi, elapsed)
}

File: pi.py
#!/usr/bin/python3
import multiprocessing as mp
import time, sys

def calc(step, base, end):
    ret = 0
    for i in range(base, end):
            x = (i + 0.5) * step
            ret = ret + 4 / (1 + x * x)
    q.put(ret)

if __name__ == "__main__":
    if sys.argv[1]:
        s = sys.argv[1]
        if s == "maxsize":
            split = sys.maxsize
        else:
            split = int(sys.argv[1])

    num_thread = int(sys.argv[2]) if len(sys.argv) > 2 else mp.cpu_count()
    step = 1.0 / split
    thread_split = split / num_thread
    _threads = []

    print("split: {0}, step: {1}".format(split, step))

    q = mp.Queue()

    start = time.time()
    for i in range(0, num_thread):
        base = int(i * thread_split)
        end = int(base + thread_split)
        _threads.append(mp.Process(target=calc, args=(step, base, end,)))
        _threads[i].start()
        print(str(i) + " start")

    for i in range(0, num_thread):
        _threads[i].join()
    end = time.time()

    sum = 0
    for i in range(0, num_thread):
        sum += q.get()

    print("pi: {0}, time: {1}".format(step*sum, end - start))
